### PR TITLE
Implement auto layout.

### PR DESCRIPTION
**Pull request by typesupply**
This addresses #4. I've implemented the API proposed in #4 and added two new wrappers around NSStackView: `HorizontalStackGroup` and `VerticalStackGroup`. These two, combined with the base API, are quite powerful. At this point, I think we should start trying this out in RoboFont experimental scripts/extensions to see how the API feels. I think that's going to be the best way to move this forward. There are still things that could be added, need documentation, etc. but that's true for all of vanilla. These changes should be completely backwards compatible, so I think this is a low risk merge.

### CODE REVIEW

**Comment by typemytype** [comment]
What mainly is missing is some examples and docs...

I would pythonize the examples from the [auto layout guide](https://developer.apple.com/library/archive/documentation/UserExperience/Conceptual/AutolayoutPG/VisualFormatLanguage.html)

and simple examples for the new StackGroups

the rest looks fine!

I would propose to build a fresh RF with this vanilla branch, so the existing ui's can be tested, next to the fresh new auto layout implementation. Merging when docs, examples are done

**Comment by typesupply** [comment]
> What mainly is missing is some examples and docs...

I added docs in the vanilla style (aka "the bare minimum") but some examples are needed. It's going to take a while because there isn't much doc out there so I'm going to have to figure it out through trial and error. I started building a UI as a test case and doc sample this way, but testing through OS X Python 2.7 is cumbersome and annoying. Plus, the vanilla docs are a mess and it made me feel overwhelmed when I tried to fix them. I'm not sure if I should add to that mess, start something new or just move on to other work.

*The visual format language reference is pretty incomplete. In fact, it's kind of odd how incomplete it is. It makes me wonder if Apple is planning on deprecating it. If they do, we can write a parser and/or take a dict for `posSize` and create the NSLayoutConstraint with [this](https://developer.apple.com/documentation/uikit/nslayoutconstraint/1526954-constraintwithitem?language=objc).

**Comment by justvanrossum** [comment]
> but testing through OS X Python 2.7 is cumbersome

Should we perhaps tag the current state, and drop support for 2.7 now?

**Comment by benkiel** [comment]
I would agree with dropping Python 2.7. As for the docs: I don't think this is the time/place to fix them, so perhaps just add to the mess?

**Comment by typemytype** [comment]
Ive adjusted the example a bit so it works with this branch

https://gist.github.com/typemytype/b43cec4608c7d5f809c0cbc9f5be9a4c

**Comment by typesupply** [comment]
A few things for my to do list:

- [ ] Add support for a dict as `posSize` that could be used to create [NSLayoutConstraint](https://developer.apple.com/documentation/uikit/nslayoutconstraint/1526954-constraintwithitem?language=objc) instead of a visual format string.
- [ ] Add support for NSLayoutConstraint.
- [ ] Write a `posSize` doc page discussing frame based positioning and auto layout and how to implement each. Put lots of examples here.

**Comment by typemytype** [comment]
how would the dict represtation look like?

as far as I understand the doc its the same as an ASCII art-like visual format string

in that case it could be alternative for the `addLayoutConstrain` attribute instead of an alternative for posSize. When an ASCII string is provided or a dict representation, posSize must be `auto` in both cases

**Comment by typesupply** [comment]
> how would the dict represtation look like?

I'm not sure yet. I've been reading a lot of the docs and presentations on auto layout and there are some things that can't be don't with the visual layout format strings. There are other times when building NSLayoutConstrain programmatically is easier to read than the strings. So, this checkbox is here for me to do research into seeing if it can be abstracted with something like the column descriptions we use in List. This would be in addition to the `(x, y, w, h)` and string representations that we now support. I'm not sure if it will work, but I want to check.

---

FWIW, I read one set of slides (from an early WWDC presentation on auto layout) that said that the recommended ways to build constraints are in this order:

1. Interface Builder
2. Visual format language strings.
3. Manually with NSLayoutConstraint.

We, obviously, don't have InterfaceBuilder. The strings are going to cover almost all situations needed by users of vanilla. So, the dict representation isn't a high priority, but I figure that it wouldn't hurt to build support for it while I understand the NSLayoutConstraint mechanics.

**Comment by typesupply** [comment]
Ugh. I just realized why my "dict as posSize" note was confusing. I meant allow a dict in `addLayoutConstraint`. Sorry! I have too many things on my mind right now.

---

My to do list:

- [ ] Add support for a dict as constraints in `addLayoutConstraints` that could be used to create [NSLayoutConstraint](https://developer.apple.com/documentation/uikit/nslayoutconstraint/1526954-constraintwithitem?language=objc) instead of a visual format string.
- [ ] Add support for [NSLayoutGuide](https://developer.apple.com/documentation/appkit/nslayoutguide?language=objc).
- [ ] Write a `posSize` doc page discussing frame based positioning and auto layout and how to implement each. Put lots of examples here.

**Comment by typemytype** [comment]
a proposal and working example of a nsLayoutAttribute as dict (requires this branch)

https://gist.github.com/typemytype/e3771d87dde7b1837c9d91ef9283fdd6

**Comment by typemytype** [comment]
(ps: this example is already something you cannot do with normal posSize settings...)

**Comment by typesupply** [comment]
> a proposal and working example of a nsLayoutAttribute as dict (requires this branch)

Yeah. This is what my vague idea was about. I'll add this and document it once I am done fighting with a typeface that doesn't want to look the way I want it to.

**Comment by typesupply** [comment]
I'm wondering if we should change the name of `addLayoutConstraints` to `addAutoPosSizeRules` to be more in line with terminology we use elsewhere like "auto" and "posSize". We don't use "layout" or "constraints" anywhere.

**Comment by typemytype** [comment]
`addAutoPosSizeRules` is fine

**Comment by typemytype** [comment]
with the latest commit I get a nasty:

```traceback
vanilla/vanillaBase.py", line 496, in _breakCycles
    obj = view.vanillaWrapper()
RecursionError: maximum recursion depth exceeded
```

File: Lib/vanilla/vanillaBox.py [hunk 1]
```diff
@@ -99,6 +99,14 @@ def __init__(self, posSize):
...
         self._nsObject.setTitlePosition_(NSNoTitle)
 
+    def _getContentView(self):
+        return self._nsObject
+
+    def _breakCycles(self):
+        view = self._getContentView()
+        if view is not None:
+            _breakCycles(view)
+
```
Comment (typemytype) [inline Lib/vanilla/vanillaBox.py]: there is already a `_breakCycles` in the Box class, this is causing a RecursionError

Comment (typesupply) [reply Lib/vanilla/vanillaBox.py]: Yeah, I see it. The root of the problem is that `NSBox` with a type `NSBoxSeparator`, which is what we do in lines, doesn't have a `contentView`. When `_getContentView` is called by vanilla's auto layout system, it gets `None` and everything breaks. I overrode `_getContentView` in `_Line` and that caused a problem with `_breakCycles`. I'm working on it.

Comment (typesupply) [reply Lib/vanilla/vanillaBox.py]: This shuld be fixed with d10098e.

**Comment by benkiel** [comment]
@typesupply Should be merge this?

**Comment by typesupply** [comment]
I think it's okay to merge. I've been using it a lot.

**Comment by benkiel** [comment]
Done!

**Pull request merged**